Fix remote build cache support in Hilt Gradle Plugin

remcomokveld opened the pull request:
Fixes https://github.com/google/dagger/issues/2763

~As proposed in the reported issue, this PR changes the annotations on the `additionalClassesDir` to match what Gradle would expect when referencing a directory that's within the root directory of the project.~ This was changed to mark it as `@Internal` to make sure the field is ignored in the cache-key calculation. 

I added a unit test which, when run on the current master fails because the `:transformDebugClassesWithAsm` results in a cache miss. On this branch it does succeed.

runningcode commented on java/dagger/hilt/android/plugin/src/test/kotlin/BuildCacheTest.kt:
```diff
@@ -0,0 +1,112 @@
+      ":extractDeepLinksDebug",
+      ":generateDebugBuildConfig",
+      ":generateDebugResValues",
+      ":javaPreCompileDebug",
```
should `javaCompile` also be in here?

remcomokveld replied:
At this point it doesn't come from cache in the second run, so would need to check why that is.

runningcode replied:
Another strategy is to include a list of all tasks in the build like we do in the Andriod cache fix plugin: https://github.com/gradle/android-cache-fix-gradle-plugin/blob/master/src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy#L540

If any tasks are added to the build they must be explicitly added there.

remcomokveld replied:
> should `javaCompile` also be in here?

Oh its actually there, it is just named `compileDebugJavaWithJavac` not `javaCompile`

runningcode reviewed (approved):
Thank you very much for including a cache relocateability test here!

remcomokveld commented:
While trying this out on a real project I found that the new annotations cause failures when a project doesn't actually have any entry points to transform

runningcode commented:
Can you mark it as `@Optional`?

remcomokveld commented:
> Can you mark it as `@Optional`?

Just tried that, and it seems to only work when the project uses javac as annotation processor. I think I have uncovered a bigger problem here which means that the logic in the method linked below only works when javac is used and not with kapt.
https://github.com/google/dagger/blob/2ba9ce05354f1153dd12905515342b8d20ea3885/java/dagger/hilt/android/plugin/src/main/kotlin/dagger/hilt/android/plugin/AndroidEntryPointClassVisitor.kt#L185-L189

remcomokveld commented:
But I feel like the fact that `@Optional` is not working here is not a bug in Dagger, but rather a bug in Gradle. The `DefaultFinalizingValidatingProperty` still does have `optional` set to true, but it still fails because `javac` didn't create the directory. I would expect Gradle's property validation nog to fail in this case.

remcomokveld commented:
So it turns out that the issue was not related to kapt, but the project I was trying this out on did not generate any code at all, which is only the case if BuildConfig generation is disabled.

I've added a test for that use case now and added a workaround for what I believe is incorrect Gradle behavior by using a FileCollection instead of of references to the one directory.

danysantiago commented on java/dagger/hilt/android/plugin/src/main/kotlin/dagger/hilt/android/plugin/AndroidEntryPointClassVisitor.kt:
```diff
@@ -207,10 +213,10 @@ class AndroidEntryPointClassVisitor(
-  private fun findAdditionalClassFile(className: String) =
-    File(additionalClasses, "$className.class")
+  private fun findAdditionalClassFile(className: String): File? =
+    additionalClasses.takeIf { it.count() == 1 }?.singleFile?.let { File(it, "$className.class") }
```
Since the use of `FileCollection` is a workaround but when actually looking into the directory we expect a single directory (file) in the collection, should we then just use `.singleFile()` without the `takeIf` so that it errors out if there is zero or more than one file. It would also make this method return not-null File

remcomokveld replied:
After a conversation in the Gradle community slack, I've actually come to reconsider this fix. 

`@Input`, `@InputFiles` etc. are used to determine if the task is up-to-date and if there is a cache miss. Since the visitor is part to the `TransformClassesWithAsmTask` inputs, and that class already defines the full classpath as an input property, this additional property can actually be ignored as part of the cache key by marking it as `@Internal` instead of `@File`. That way there are zero behavior changes in this visitor itself, but the relocatability test still succeeds.

danysantiago replied:
Awesome, thanks for the update.

danysantiago reviewed (commented):
Thanks for sending this PR!

remcomokveld commented:
Let me know if you prefer to keep the history in this PR, or rather rebase it to remove the intermediate commits.

danysantiago commented:
No need to rebase, the bot that merges things will do that.
